### DISCUSSION TITLE
Integrated Metrics

### DISCUSSION THREAD

**Discussion by danielrehmann** [tomorrow-one/transactional-outbox#135] [Ideas]
Recently there was an issue with a `OutboxProcessor` not working in one of our services. Currently we don't have any metrics regarding our outbox libraries and offset in our outbox table.

I was wondering if we could introduce meter registry or something that allows us to add metrics like outbox table offset.

I have to admit I am unsure if this should be part of this library or projects need to implement it themselves so I wanted to open a discussion first.

**Comment by magro** [comment]
I definitely like the idea of instrumentation / observability improvements. Maybe we collect some metrics that would be good to have? E.g.
* `num_processed_records` (the number of `OutboxRecord` entries that have set `processed`)
* `num_unprocessed_records` (the number of `OutboxRecord` entries that have _not_ set `processed`)

?

**Reply by danielrehmann** [reply]
I think for me the `num_unprocessed_records` containing the amount of outbox entities that not have been sent yet would be essential. Thats a metric that can be easily monitored and alerted on.

`num_processed_records` is a nice to have where we could actually monitor the throughput of our Outbox.

**Comment by magro** [comment]
I'm also curious to hear more about the issue and root cause, which lead to "a `OutboxProcessor` not working". What exactly means "not working", and what was the cause for that? Is there anything in this library that could be improved to prevent this situation in the future?

**Reply by mrhnrk** [reply]
I think what @danielrehmann is talking about is the issue we had with one of our services after updating the database. The library was not able to read the unprocessed records from the database any more, if I understood it correctly. We fixed this by restarting the service. @danielrehmann, maybe you have deeper details here? 🤗

**Reply by danielrehmann** [reply]
@hnrkdmsk exactly, as it was a hibernate/jpa service managed by a different team that was affected I did in fact not have deeper insights here.

From the lessons learned of that incident the topic of metrics and alerting was raised. Currently we are quite unaware of the state of the outbox table.

Looking into further hardening might also be a good topic but this discussion had metrics and alerting in mind.

**Reply by magro** [reply]
Ok, thanks. So I guess either the interaction with the lock repository or with the outbox repository failed, do you have some stack traces maybe? And I'm wondering if writing to the outbox was still working...?

With this additional knowledge, it might make sense to have also metrics like e.g.
* num_outbox_processor_generic_failures - this might be incremented in various places where exceptions are handled

**Reply by danielrehmann** [reply]
I did not find any logs that could help us regarding this incident, maybe @hnrkdmsk could help here.

**Reply by mrhnrk** [reply]
@ChulioZet, maybe you can help out here with providing some logs? 🤗

**Reply by ghost** [reply]
Unfortunately, I also cannot find any log messages that seem helpful here. What I can say is that our database was unavailable for at least 1:45 minutes due to a minor update. After the database was available again, reading and writing to the database seemed to work flawlessly again -- for the transactional-outbox as well as for all our own services. At least I cannot find any errors related to database connections not working and the system worked "as usual". So I suspect that writing into the outbox table wasn't the issue.

Instead, it seems as if the transactional-outbox hasn't picked up anything from the outbox table anymore -- no Kafka events have been published since the database downtime and until I restarted our service. After the restart, the outbox published (what seemed to me like) all events that were waiting in the outbox table since the downtime. This, again, went through without problems. So the only issue I see is that no events have been published for those entries that have been inserted into the outbox between database downtime and service restart.

**Reply by magro** [reply]
Ok, that's definitely interesting, thanks! It could make sense to try to reproduce this in a test like `OutboxProcessorIntegrationTest`...?

**Reply by danielrehmann** [reply]
I think this issue got again more urgent, as it happened again @ChulioZet @hnrkdmsk could you maybe open an issue describing what resulted in the outbox not sending messages?

**Reply by mrhnrk** [reply]
I'm not so familiar with this topic, @ChulioZet can you take over please? 🤗

**Reply by magro** [reply]
I could reproduce the issue and just submitted a PR with the fix: https://github.com/tomorrow-one/transactional-outbox/pull/183. To make megalinter/editorconfig rules happy I had to replace tabs with spaces for the `OutboxProcessor` - if you want to see just the relevant change please activate "Hide whitespace" in the diff view settings, or see commit [61c393](https://github.com/tomorrow-one/transactional-outbox/pull/183/commits/61c393e2276f5ed664c62e94bd83f499bf617a5a).

**Reply by ghost** [reply]
Perfect, that sounds good. Then I won't open an issue or anything right now. If you need any additional info or something else, just reach out. 🙂

**Reply by magro** [reply]
If you have analyzed the issue it would still be valuable to compare the issue that I reproduced and fixed here with your analysis / observations, @ChulioZet.

**Reply by ghost** [reply]
I haven't analyzed too much yet. So I'd lean towards trusting your fix for now (if the reviewing maintainers are fine with it 🙂) until we find out that the issue arises again.
